src/OpenCLScheduler.cpp

```diff
@@ -182,28 +195,92 @@ template <typename net_t>
 void OpenCLScheduler<net_t>::forward(const std::vector<float>& input,
                                      std::vector<float>& output_pol,
                                      std::vector<float>& output_val) {
-    std::shared_ptr<ContextPoolEntry> ctx;
-    auto queue_num = size_t{0};
+    auto entry = std::make_shared<ForwardQueueEntry>(input, output_pol, output_val);
+    std::unique_lock<std::mutex> lk(entry->mutex);
     {
-        LOCK(m_context_pool_mutex, lock);
-        while (queue_num < m_context_pool.size()) {
-            if (!m_context_pool[queue_num].empty()) {
-                ctx = std::move(m_context_pool[queue_num].front());
-                m_context_pool[queue_num].pop_front();
-                break;
+        std::unique_lock<std::mutex> lk(m_mutex);
+        m_forward_queue.push_back(entry);
+    }
+    m_cv.notify_one();
+    entry->cv.wait(lk);
+}
+
+std::atomic<size_t> batch_stats[2];
+
+template <typename net_t>
+void OpenCLScheduler<net_t>::set_batching(bool is_batching) {
+    m_is_batching = is_batching;
+    m_cv.notify_one();
+}
+
+template <typename net_t>
+void OpenCLScheduler<net_t>::batch_worker(const size_t gnum) {
+    auto batch_input = std::vector<float>(Network::INPUT_CHANNELS * BOARD_SIZE * BOARD_SIZE * cfg_batch_size);
+    auto batch_output_pol = std::vector<float>(Network::OUTPUTS_POLICY * BOARD_SIZE * BOARD_SIZE * cfg_batch_size);
+    auto batch_output_val = std::vector<float>(Network::OUTPUTS_VALUE * BOARD_SIZE * BOARD_SIZE * cfg_batch_size);
+    OpenCLContext context;
+
+    while (true) {
+        std::list<std::shared_ptr<ForwardQueueEntry>> inputs;
+        size_t count = 0;
+
+        {
+            std::unique_lock<std::mutex> lk(m_mutex);
+            while (true) {
+                if(!m_running) return;
+                count = std::min(m_forward_queue.size(), size_t(cfg_batch_size));
+                if (count < cfg_batch_size && m_is_batching) {
+                    count = 0;
+                }
+                if (count > 0) {
+                    auto begin = m_forward_queue.begin();
+                    auto end = begin;
+                    std::advance(end, count);
+                    std::move(begin, end, std::back_inserter(inputs));
+                    m_forward_queue.erase(begin, end);
+                    break;
+                }
+                else {
+                    // wait with pred does not work anymore, not sure why yet
+                    m_cv.wait(lk);
+                    //m_cv.wait(lk, [this](){ return !m_running || !m_forward_queue.empty(); });
```

Comment:
It'll stuck after my full batch changes, not sure why. It was fine before my changes.
And I'm not sure why we need the predicts here, since when notified, the queue should never be empty.

Reply:
I think the inner loop already covered this, so we don't need it anyway.